kafka-connect/src/main/java/com/hazelcast/jet/contrib/connect/KafkaConnectSources.java

```diff
@@ -0,0 +1,193 @@
+
+import com.hazelcast.instance.impl.HazelcastInstanceFactory;
+import com.hazelcast.jet.JetInstance;
+import com.hazelcast.jet.core.Processor;
+import com.hazelcast.jet.pipeline.SourceBuilder;
+import com.hazelcast.jet.pipeline.SourceBuilder.TimestampedSourceBuffer;
+import com.hazelcast.jet.pipeline.StreamSource;
+import org.apache.kafka.connect.connector.ConnectorContext;
+import org.apache.kafka.connect.source.SourceConnector;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.apache.kafka.connect.source.SourceTask;
+import org.apache.kafka.connect.source.SourceTaskContext;
+import org.apache.kafka.connect.storage.OffsetStorageReader;
+
+import java.net.URL;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import static com.hazelcast.jet.impl.util.ExceptionUtil.rethrow;
+
+/**
+ * Contains factory methods to create a Kafka Connect source.
+ */
+public final class KafkaConnectSources {
+
+    private KafkaConnectSources() {
+    }
+
+    /**
+     * A generic Kafka Connect source provides ability to plug any Kafka Connect
+     * source for data ingestion to Jet pipelines.
+     * <p>
+     * You need to add the Kafka Connect connector JARs or a ZIP file contains
+     * the JARs as a job resource via {@link com.hazelcast.jet.config.JobConfig#addJar(URL)}
+     * or {@link com.hazelcast.jet.config.JobConfig#addJarsInZip(URL)} respectively.
+     * <p>
+     * After that you can use the Kafka Connect connector with the configuration
+     * parameters as you'd using it with Kafka. Hazelcast Jet will drive the
+     * Kafka Connect connector from the pipeline and the records will be available
+     * to your pipeline as {@link SourceRecord}s.
+     * <p>
+     * In case of a failure; this source keeps track of the source partition
+     * offsets, it will restore the partition offsets and resume the consumption
+     * from where it left off.
+     *
+     * @param properties Kafka connect properties
+     * @return a source to use in {@link com.hazelcast.jet.pipeline.Pipeline#readFrom(StreamSource)}
+     */
+    public static StreamSource<SourceRecord> connect(Properties properties) {
+        String name = properties.getProperty("name");
+        return SourceBuilder.timestampedStream(name, ctx -> new Context(ctx, properties))
+                            .fillBufferFn(Context::fillBuffer)
+                            .createSnapshotFn(Context::createSnapshot)
+                            .restoreSnapshotFn(Context::restoreSnapshot)
+                            .destroyFn(Context::destroy)
+                            .build();
+    }
+
+    private static class Context {
+
+        private final SourceConnector connector;
+        private final SourceTask task;
+        private final Map<String, String> taskConfig;
+
+        private Map<Map<String, ?>, Map<String, ?>> partitionsToOffset = new HashMap<>();
```

Comment:
can we have a javadoc describes what does this map contains exactly ? I'm a bit confused, is a partition represented as a map? and what dose that map contains ?